### PR TITLE
MOD Switch class variables to constant

### PR DESCRIPTION
**Pull request by Guillermo849**

### CODE REVIEW

File: caesar_cipher.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            new_letter_value += num_change
+            # Checks if the index number is within the length of the alphabet, if it's not it will start form the start of the alphabet or from the end
+            if new_letter_value > fill_alphabet_lower.length()-1
+                new_letter_value = 0
+            end
+            if new_letter_value < 0
+                new_letter_value = fill_alphabet_lower.length()-1
+            end
+        end
+        return fill_alphabet_lower.key(new_letter_value).to_s
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:72]: In Ruby, methods ``return`` the value of the last evaluated expression by default, so in this case you do not need to use ``return``
```suggestion
        fill_alphabet_lower.key(new_letter_value).to_s
```

File: caesar_cipher.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            new_letter_value += num_change
+            # Checks if the index number is within the length of the alphabet, if it's not it will start form the start of the alphabet or from the end
+            if new_letter_value > fill_alphabet_upper.length()-1
+                new_letter_value = 0
+            end
+            if new_letter_value < 0
+                new_letter_value = fill_alphabet_upper.length()-1
+            end
+        end
+        return fill_alphabet_upper.key(new_letter_value).to_s
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:89]: Same with ``return``
```suggestion
       fill_alphabet_upper.key(new_letter_value).to_s
```

File: caesar_cipher.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    end
+
+    private
+
+    attr_reader :text, :num
+    
+    # Will create a hash for the uppercase letter or the lower case ones
+    def fill_alphabet_upper
+        return @fill_alphabet_upper if defined? @fill_alphabet_upper
+        @fill_alphabet_upper = Hash.new
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:27]: You can just use  `{}`
 ```suggestion
        @fill_alphabet_upper = {}
```

File: caesar_cipher.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        ALPHABET_UPPERCASE.each_with_index do |key, l|
+            @fill_alphabet_upper[key.to_sym] = l
+        end
+        @fill_alphabet_upper
+    end
+
+    def fill_alphabet_lower
+        return @fill_alphabet_lower if defined? @fill_alphabet_lower
+        @fill_alphabet_lower = Hash.new
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:36]:
```suggestion
        @fill_alphabet_lower = {}
```

File: caesar_cipher.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        @text = text
+        @num = num
+    end
+
+    # Method that ciphers the text that we give it
+    def cipher
+        # It will check if there is text to cipher and if there is a valid number to change the letters
+        return puts "The text given is null and/or the number given is 0." if text == "" || num.zero?
+        text_split = text.split('')
+        puts refactor_text(text_split)
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:17]: To make the code more readable, it is worth adding one blank line between the `return` method and the business logic.  
```suggestion
        return puts "The text given is null and/or the number given is 0." if text == "" || num.zero?
        
        text_split = text.split('')
        puts refactor_text(text_split)
```

File: caesar_cipher.rb [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    def initialize(text, num)
+        @text = text
+        @num = num
+    end
+
+    # Method that ciphers the text that we give it
+    def cipher
+        # It will check if there is text to cipher and if there is a valid number to change the letters
+        return puts "The text given is null and/or the number given is 0." if text == "" || num.zero?
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:15]: What will the program return if the user passes an blank string ``' '``? 
IMO, you should check the case where the string is blank. 
```suggestion
        return puts "The text given is null and/or the number given is 0." if text.strip.empty? || num.zero?
```
Please check how it works in your ruby console(`irb`) 
DOCS for strip method: https://reintech.io/blog/exploring-the-string-strip-method-in-ruby

File: caesar_cipher.rb [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            @fill_alphabet_upper[key.to_sym] = l
+        end
+        @fill_alphabet_upper
+    end
+
+    def fill_alphabet_lower
+        return @fill_alphabet_lower if defined? @fill_alphabet_lower
+        @fill_alphabet_lower = Hash.new
+        ALPHABET_LOWERCASE.each_with_index do |key, l|
+            @fill_alphabet_lower[key.to_sym] = l
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:38]: IMO, converting a string to a symbol further complicates your code, try to use string for key in hash.
```suggestion
            @fill_alphabet_lower[key] = l
```

File: caesar_cipher.rb [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            end
+            if new_letter_value < 0
+                new_letter_value = fill_alphabet_lower.length()-1
+            end
+        end
+        return fill_alphabet_lower.key(new_letter_value).to_s
+    end
+
+    # Private method that transforms uppercase letters
+    def uppper_letter(letter)
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:76]: This method contains the same logic as `lower_letter`, so you can refactor them. Try to  use one method but with different collections.
DRY -> https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

**Review by OlekKovalevych — CHANGES_REQUESTED** [review]

File: caesar_cipher.rb [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    end
+
+    # Private method that transforms lowercase letters
+    def lower_letter(letter)
+        new_letter_value = fill_alphabet_lower[letter].to_i
+        num_change = num < 0? -1 : 1
+        for v in 1..num.abs do
+            new_letter_value += num_change
+            # Checks if the index number is within the length of the alphabet, if it's not it will start form the start of the alphabet or from the end
+            if new_letter_value > fill_alphabet_lower.length()-1
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:65]: Could you please fix the code style in the whole file
```suggestion
            if new_letter_value > fill_alphabet_lower.length - 1
```

File: caesar_cipher.rb [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                new_letter_value = 0
+            end
+            if new_letter_value < 0
+                new_letter_value = fill_alphabet_lower.length()-1
+            end
```
Comment (OlekKovalevych) [inline caesar_cipher.rb:70]: Could you please convert this statement  to Single-line Conditional, please make also in other cases where it is possible

File: caesar_cipher.rb (hunk 9 above)
Comment (OlekKovalevych) [inline caesar_cipher.rb:65]: Maybe in this way? 
```suggestion
            if new_letter_value >= fill_alphabet_lower.length
```

File: caesar_cipher.rb (hunk 3 above)
Comment (Guillermo849) [reply caesar_cipher.rb]: +1

File: caesar_cipher.rb (hunk 5 above)
Comment (Guillermo849) [reply caesar_cipher.rb]: +1

File: caesar_cipher.rb (hunk 7 above)
Comment (Guillermo849) [reply caesar_cipher.rb]: +1

File: caesar_cipher.rb (hunk 6 above)
Comment (Guillermo849) [reply caesar_cipher.rb]: +1

File: caesar_cipher.rb (hunk 8 above)
Comment (Guillermo849) [reply caesar_cipher.rb:73]: +1

File: caesar_cipher.rb (hunk 10 above)
Comment (Guillermo849) [reply caesar_cipher.rb]: +1

File: caesar_cipher.rb (hunk 9 above)
Comment (Guillermo849) [reply caesar_cipher.rb]: +1

**Review by OlekKovalevych — APPROVED** [review]

**Pull request merged**